File: programmers/거리두기 확인하기.py
dx = [1, -1, 0, 0]
dy = [0, 0, 1, -1]
    
def BST(x, y, room):
    visited = [[0 for _ in range(5)] for _ in range(5)]
    man_len = 0
    tmp1 = [[x, y]]
    tmp2 = [0]
    
    while len(tmp1) > 0:
        x = tmp1[0][0]
        y = tmp1[0][1]
        tmp1.remove(tmp1[0])
        visited[x][y] = 1
        man_len = tmp2[0]
        tmp2.remove(tmp2[0])
        
        if man_len == 2: # 맨해튼 거리 2 초과이므로 더 이상 진행할 필요 없음
            return 1
        
        for i in range(4):
            nx = x + dx[i]
            ny = y + dy[i]
            if 0 <= nx < 5 and 0 <= ny < 5:
                if visited[nx][ny] == 0:
                    if room[nx][ny] == 'P':
                        return 0
                    elif room[nx][ny] == 'O':
                        tmp1.append([nx, ny])
                        tmp2.append(man_len + 1)
    return 1

def solution(places):
    answer = []
    for room in places:
        stop = False
        for x in range(5):
            for y in range(5):
                if room[x][y] == 'P':
                    if BST(x, y, room) == 0:
                        stop = True
                        break
            if stop == True:
                break
        if stop == True:
            answer.append(0)
        else:
            answer.append(1)
    return answer

File: programmers/level2/거리두기_확인하기.cpp
#include <string>
#include <vector>

using namespace std;

int dx[4] = { 0,0,-1,1 }; // 상하좌우
int dy[4] = { 1,-1,0,0 };

int check(const vector<string> &room) {
    for (int x=0; x<5; ++x) {
        for (int y=0; y<5; ++y) {
            if (room[x][y] == 'P') { // 응시자 확인
                for (int i=0; i<4; ++i) { // 응시자 상하좌우 (맨허튼 거리 1)
                    int nx = x + dx[i];
                    int ny = y + dy[i];
                    if (nx<0 || ny<0 || nx>=5 || ny>=5) // 경계
                        continue;
                    if (room[nx][ny] == 'X') // 파티션
                        continue;
                    if (room[nx][ny] == 'P') // 거리두기 X
                        return 0;
                    for (int j=0; j<4; ++j) { // 맨허튼 거리 2
                        int nnx = nx + dx[i];
                        int nny = ny + dy[i];
                        if (nnx==x && nny==y) // 자기자신 이미 확인
                            continue;
                        if (nnx<0 || nny<0 || nnx>=5 || nny>=5)
                            continue;
                        if (room[nnx][nny] == 'X')
                            continue;
                        if (room[nnx][nny] == 'P')
                            return 0;
                    }
                }
            }
        }
    }
    return 1;
}
vector<int> solution(vector<vector<string>> places) {
    vector<int> answer;
    for (auto room : places) {
        answer.push_back(check(room));
    }
    return answer;
}
// 합계: 82.8 / 100.0